File: tools/EventClients/examples/c++/example_button1.cpp
#include "../../lib/c++/xbmcclient.h"
#include <stdio.h>
#include <string.h>
#include <sys/socket.h>
#ifdef _WIN32
#include <Windows.h> // for sleep
#else
#include <unistd.h>
#endif

int main(int argc, char **argv)
{
  /* connect to localhost, port 9777 using a UDP socket
     this only needs to be done once.
     by default this is where XBMC will be listening for incoming
     connections. */
  CAddress my_addr; // Address => localhost on 9777
  int sockfd = socket(AF_INET, SOCK_DGRAM, 0);
  if (sockfd < 0)
  {
    printf("Error creating socket\n");
    return -1;
  }

  my_addr.Bind(sockfd);

  std::string sIconFile = "../../icons/bluetooth.png";
  unsigned short usIconType = ICON_PNG;

  std::ifstream file (sIconFile, std::ios::in|std::ios::binary|std::ios::ate);
  if (!file.is_open())
  {
    sIconFile = "/usr/share/pixmaps/kodi/bluetooth.png";
    file.open(sIconFile, std::ios::in|std::ios::binary|std::ios::ate);

    if (!file.is_open()) {
      usIconType = ICON_NONE;
    }
    else
    {
      file.close();
    }
  }
  else
  {
    file.close();
  }

  CPacketHELO HeloPackage("Example Remote", usIconType, sIconFile.c_str());
  HeloPackage.Send(sockfd, my_addr);

  sleep(5);
  // press 'S'
  CPacketBUTTON btn1('S', true);
  btn1.Send(sockfd, my_addr);

  sleep(2);
  // press the enter key (13 = enter)
  CPacketBUTTON btn2(13, true);
  btn2.Send(sockfd, my_addr);

  // BYE is not required since XBMC would have shut down
  CPacketBYE bye; // CPacketPing if you want to ping
  bye.Send(sockfd, my_addr);
}

File: tools/EventClients/examples/python/example_button1.py
#!/usr/bin/python

# This is a simple example showing how you can send 2 button events
# to XBMC in a queued fashion to shut it down.

# Queued button events are not repeatable.

# The basic idea is to create single packets and shoot them to XBMC
# The provided library implements some of the support commands and
# takes care of creating the actual packet. Using it is as simple
# as creating an object with the required constructor arguments and
# sending it through a socket.

# Currently, only keyboard keys are supported so the key codes used
# below are the same key codes used in guilib/common/SDLKeyboard.cpp

# In effect, anything that can be done with the keyboard can be done
# using the event client.

# import the XBMC client library
# NOTE: The library is not complete yet but is usable at this stage.

import sys
sys.path.append("../../lib/python")

from xbmcclient import *
from socket import *

def main():
    import time
    import sys

    # connect to localhost, port 9777 using a UDP socket
    # this only needs to be done once.
    # by default this is where XBMC will be listening for incoming
    # connections.
    host = "localhost"
    port = 9777
    addr = (host, port)
    sock = socket(AF_INET,SOCK_DGRAM)

    # First packet must be HELO (no it's not a typo) and can contain an icon
    # 'icon_type' can be one of ICON_NONE, ICON_PNG, ICON_JPG or ICON_GIF
    packet = PacketHELO(devicename="Example Remote",
                        icon_type=ICON_PNG,
                        icon_file="../../icons/bluetooth.png")
    packet.send(sock, addr)

    # IMPORTANT: After a HELO packet is sent, the client needs to "ping" XBMC
    # at least once every 60 seconds or else the client will time out.
    # Every valid packet sent to XBMC acts as a ping, however if no valid
    # packets NEED to be sent (eg. the user hasn't pressed a key in 50 seconds)
    # then you can use the PacketPING class to send a ping packet (which is
    # basically just an empty packet). See below.

    # Once a client times out, it will need to reissue the HELO packet.
    # Currently, since this is a unidirectional protocol, there is no way
    # for the client to know if it has timed out.

    # wait for notification window to close (in XBMC)
    time.sleep(5)

    # press 'S'
    packet = PacketBUTTON(code='S', queue=1)
    packet.send(sock, addr)

    # wait for a few seconds
    time.sleep(2)

    # press the enter key (13 = enter)
    packet = PacketBUTTON(code=13, queue=1)
    packet.send(sock, addr)

    # BYE is not required since XBMC would have shut down
    packet = PacketBYE()    # PacketPING if you want to ping
    packet.send(sock, addr)

if __name__=="__main__":
    main()
